pkg/controller/azvolumeattachment.go

```diff
@@ -0,0 +1,546 @@
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// If it is the first time seeing this underlying volume, sync scMap and volMap
+		if _, ok := r.volMap[azVolumeAttachment.Spec.UnderlyingVolume]; !ok {
+			if err := r.SyncStorageClass(ctx); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+
+		// update the status
+		updated := azVolumeAttachment.DeepCopy()
+		updated.Status = &v1alpha1.AzVolumeAttachmentStatus{}
+		updated.Status.AttachmentTier = updated.Spec.AttachmentTier
+		updated.Status.AttachmentState = v1alpha1.Attached
+
+		if azVolumeAttachment.Spec.AttachmentTier == v1alpha1.Primary {
+			primary, primaryOk := r.primaryMap[azVolumeAttachment.Spec.UnderlyingVolume]
+			// this is a case when previous primary attachment object had not been deleted for some reason,
+			// in this case, mark the old primary detach required, and update the primary map
+			if primaryOk && primary.Name != azVolumeAttachment.Name {
+				if err := r.TriggerDetach(ctx, primary); err != nil {
+					// if detach failed, requeue
+					return reconcile.Result{Requeue: true}, err
+				}
+				// this is a case when the attachment object is created for the first time for the specified underlying volume
+				// in this case, create n replicas
+			} else {
+				if err := r.ManageReplicaForVolume(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err != nil {
+					return reconcile.Result{Requeue: true}, err
+				}
+			}
+		}
+
+		if _, err := r.azVolumeClient.DiskV1alpha1().AzVolumeAttachments(r.namespace).UpdateStatus(ctx, updated, metav1.UpdateOptions{}); err != nil {
+			klog.Errorf("failed to update status for azvolumeattachment object %s: (%v)", azVolumeAttachment.Name, err)
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// if status update was successful add the object to either the replica or primary map
+		if azVolumeAttachment.Spec.AttachmentTier == v1alpha1.Primary {
+			r.primaryMap[azVolumeAttachment.Spec.UnderlyingVolume] = azVolumeAttachment
+		} else {
+			r.replicaMap[azVolumeAttachment.Spec.UnderlyingVolume][azVolumeAttachment.Spec.NodeName] = azVolumeAttachment
+		}
+
+		// If it is the first time seeing this node, sync node map
+		if _, ok := r.nodeMap[azVolumeAttachment.Spec.NodeName]; !ok {
+			r.nodeMap[azVolumeAttachment.Spec.NodeName] = make(map[string]v1alpha1.AzVolumeAttachment)
+		}
+		r.nodeMap[azVolumeAttachment.Spec.NodeName][azVolumeAttachment.Spec.UnderlyingVolume] = azVolumeAttachment
+		klog.Infof("successfully attached volume %s to node %s and update %s's status", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName, azVolumeAttachment.Name)
+		// if the azVolumeAttachment's attachment state is detach required, detach the disk from the node and delete the object
+	} else if azVolumeAttachment.Status.AttachmentState == v1alpha1.DetachRequired {
```

Comment:
This should be handled through a [`finalizer`](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#finalizers) rather than through a `DetachRequired` state.

Reply:
Good point. I forgot about the `finalizers` as I was kind of swept away by the idea of dividing down possible scenarios with `AttachmentState` and reconciling.